crates/ra_project_model/src/lib.rs

```diff
@@ -25,15 +25,35 @@ pub use crate::{
 };
 
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
-pub struct CargoTomlNotFoundError(pub PathBuf);
+pub struct CargoTomlNoneFoundError(pub PathBuf);
 
-impl std::fmt::Display for CargoTomlNotFoundError {
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct CargoTomlMultipleValidFoundError(pub Vec<PathBuf>);
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct CargoTomlSearchFileSystemError(pub PathBuf, pub String);
+
+impl std::fmt::Display for CargoTomlNoneFoundError {
     fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         write!(fmt, "can't find Cargo.toml at {}", self.0.display())
     }
 }
 
-impl Error for CargoTomlNotFoundError {}
+impl std::fmt::Display for CargoTomlMultipleValidFoundError {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(fmt, "found multiple valid Cargo.toml files {:?}", self.0)
+    }
+}
+
+impl std::fmt::Display for CargoTomlSearchFileSystemError {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(fmt, "a filesystem error occurred while searching for Cargo.toml: {}", self.1)
+    }
+}
+
+impl Error for CargoTomlNoneFoundError {}
+impl Error for CargoTomlMultipleValidFoundError {}
+impl Error for CargoTomlSearchFileSystemError {}
```

Comment:
Hope this method of grouping is acceptable? Group related structs and impls together.

I don't have a strong preference, I can more them around if that is preferred. :)